segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,453 @@
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.info("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            retVal.completeExceptionally(new IllegalArgumentException());
+            return;
+        }
+
+        if (!Files.exists(path)) {
+            retVal.completeExceptionally(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+            return;
+        }
+
+        try {
+            if (!isWritableFile(path)) {
+                retVal.completeExceptionally(new StreamSegmentSealedException(handle.getSegmentName()));
+                return;
+            }
+
+            long fileSize = path.toFile().length();
+            if (fileSize < offset) {
+                retVal.completeExceptionally(new BadOffsetException(handle.getSegmentName(), fileSize, offset));
+            } else {
+                try (FileChannel channel = FileChannel.open(path,
+                        StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    channel.force(true);
+                }
+                retVal.complete(null);
+            }
+        } catch (Exception exc) {
+            log.info("Write to segment {} at offset {} failed with exception {} ", handle.getSegmentName(), offset,
+                    exc.getMessage());
+            if (exc instanceof AccessDeniedException) {
+                retVal.completeExceptionally(new IllegalStateException(handle.getSegmentName()));
+            } else if (exc instanceof NonWritableChannelException) {
+                retVal.completeExceptionally(new IllegalArgumentException(exc));
+            } else if (exc instanceof ClosedChannelException) {
+                retVal.completeExceptionally(new StreamSegmentSealedException(handle.getSegmentName(), exc));
+            } else {
+                retVal.completeExceptionally(exc);
+            }
+        }
+
+    }
+
+    private boolean isWritableFile(Path path) throws IOException {
+        PosixFileAttributes attrs = null;
+            attrs = Files.readAttributes(path,
+                    PosixFileAttributes.class);
+            return attrs.permissions().contains(OWNER_WRITE);
+
+    }
+
+
+    private void syncSeal(SegmentHandle handle, Duration timeout, CompletableFuture<Void> retVal) {
+
+        if (handle.isReadOnly()) {
+            log.info("Seal called on a read handle for segment {}", handle.getSegmentName());
+            retVal.completeExceptionally(new IllegalArgumentException(handle.getSegmentName()));
+            return;
+        }
+
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
```

Comment:
It should be documented in our architecture docs that this is what we do for sealing.